File: 0102-binary-tree-level-order-traversal/solution.cpp

class Solution {
public:
    int level(TreeNode* root) {
        if (root == nullptr)
            return 0;
        return 1 + max(level(root->left), level(root->right));
    }
    void fun(TreeNode* root, vector<vector<int>>& v,
             int curr) // left ORDER displaypre
    {
        if (root == nullptr)
            return;
        v[curr].push_back(root->val);
        fun(root->left, v, curr + 1);
        fun(root->right, v, curr + 1);
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> v;
        for (int i = 1; i <= level(root); i++) {
            vector<int> w;
            v.push_back(w);
        }
        fun(root, v, 0);
        return v;
    }
};

File: 0102-binary-tree-level-order-traversal/solution.java
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;  // edge case

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);

        while (!q.isEmpty()) {
            int size = q.size();              // current level ka size
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                TreeNode curr = q.poll();
                level.add(curr.val);

                if (curr.left != null)  q.offer(curr.left);
                if (curr.right != null) q.offer(curr.right);
            }
            result.add(level);
        }
        return result;
    }
}
